forcefield_utilities/gmso_xml.py

```diff
@@ -0,0 +1,868 @@
+        description="Class 3 for this Dihedral/Improper type",
+        alias="class3",
+    )
+
+    class4: Optional[str] = Field(
+        None,
+        description="Class 4 for this Dihedral/Improper type",
+        alias="class4",
+    )
+
+    type1: Optional[str] = Field(
+        None,
+        description="Type 1 for this Dihedral/Improper type",
+        alias="type1",
+    )
+
+    type2: Optional[str] = Field(
+        None,
+        description="Type 2 for this Dihedral/Improper type",
+        alias="type2",
+    )
+
+    type3: Optional[str] = Field(
+        None,
+        description="Type 3 for this Dihedral/Improper type",
+        alias="type3",
+    )
+
+    type4: Optional[str] = Field(
+        None,
+        description="Type 4 for this Dihedral/Improper type",
+        alias="type4",
+    )
+
+    children: List[Parameters] = Field(
+        ..., description="The parameters and their values", alias="children"
+    )
+
+    @classmethod
+    def load_from_etree(cls, root):
+        attribs = root.attrib
+        children = []
+        for el in root.iterchildren():
+            if el.tag == "Parameters":
+                children.append(Parameters.load_from_etree(el))
+        return cls(children=children, **root.attrib)
+
+
+class DihedralType(TorsionType):
+    pass
+
+
+class ImproperType(TorsionType):
+    pass
+
+
+class TorsionTypes(GMSOXMLChild):
+    name: Optional[str] = Field(
+        None, description="The name for this angle types group", alias="name"
+    )
+
+    expression: str = Field(
+        ..., description="The expression for this angle types group"
+    )
+
+    children: List[Union[ParametersUnitDef, TorsionType]] = Field(
+        ...,
+        description="Children of this dihedral/improper types tag",
+        alias="children",
+    )
+
+    def to_gmso_potentials(self, default_units):
+        potentials = {"dihedral_types": {}, "improper_types": {}}
+        parameters_units = filter(
+            lambda c: isinstance(c, ParametersUnitDef), self.children
+        )
+        units = {
+            parameter_unit.parameter: u.Unit(parameter_unit.unit)
+            for parameter_unit in parameters_units
+        }
+
+        for torsion_type in filter(
+            lambda c: isinstance(c, (DihedralType, ImproperType)), self.children
+        ):
+            torsion_dict = torsion_type.dict(
+                by_alias=True,
+                exclude={
+                    "children",
+                    "type1",
+                    "type2",
+                    "type3",
+                    "type4",
+                    "class1",
+                    "class2",
+                    "class3",
+                    "class4",
+                },
+                exclude_none=True,
+            )
+
+            if "expression" not in torsion_dict:
+                torsion_dict["expression"] = self.expression
+
+            if (
+                torsion_type.type1
+                and torsion_type.type2
```

Comment:
This doesn't work if the type or class is only partially defined e.g.
```xml
<DihedralType name="DihedralType1" type1="" type2="C" type3="C" type4="">
```

We need to switch the `and`'s for `or`'s here to check if any one is defined.

Reply:
@CalCraven Good catch. I think we need an alternative approach where in the default value is populated as a `*` and this check should stay as is.